src/query/vacuum.c

```diff
@@ -5566,6 +5583,13 @@ vacuum_update_keep_from_log_pageid (THREAD_ENTRY * thread_p)
 
   vacuum_er_log (VACUUM_ER_LOG_VACUUM_DATA,
 		 "Update keep_from_log_pageid to %lld ", (long long int) vacuum_Data.keep_from_log_pageid);
+
+  if (!vacuum_Data.is_archive_removal_safe)
+    {
+      /* remove archives that have been blocked up to this point. */
+      vacuum_Data.is_archive_removal_safe = true;
+      logpb_remove_archive_logs_exceed_limit (thread_p, 0);
```

Comment:
What do you think of moving `logpb_remove_archive_logs_exceed_limit` call to `vacuum_data_load_and_recover`?
The patch lets `vacuum_data_mark_finished` try to remove archives too.

Reply:
The archives may be removed only the first time keep_from_log_pageid is updated. vacuum_data_mark_finished would not enter the if condition since vacuum_Data.is_archive_removal_safe.

I initially added the archive removal into `vacuum_data_load_and_recover`, but then I reconsidered to tie it directly to setting the `vacuum_Data.keep_from_log_pageid` field. If in the future we change the vacuum data design and `vacuum_Data.keep_from_log_pageid` can be determined before loading, maybe archives could be removed earlier too.

However, removing the archives on loading is a valid option too, if you think it is more clear with current design.

Reply:
You are right. I agree with the current :)